flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferPersisterImpl.java

```diff
@@ -0,0 +1,260 @@
+		private final RecoverableWriter recoverableWriter;
+		private final Path recoverableWriterBasePath;
+
+		@Nullable
+		private Throwable asyncException;
+		private int partId;
+		private CompletableFuture<?> persistFuture = CompletableFuture.completedFuture(null);
+		private RecoverableFsDataOutputStream currentOutputStream;
+		private byte[] readBuffer = new byte[0];
+
+		public Writer(RecoverableWriter recoverableWriter, Path recoverableWriterBasePath) throws IOException {
+			this.recoverableWriter = recoverableWriter;
+			this.recoverableWriterBasePath = recoverableWriterBasePath;
+
+			openNewOutputStream();
+		}
+
+		public synchronized void add(Buffer buffer) {
+			checkErroneousUnsafe();
+			boolean wasEmpty = handover.isEmpty();
+			handover.add(buffer);
+			if (wasEmpty) {
+				notify();
+			}
+		}
+
+		public synchronized CompletableFuture<?> persist() {
+			checkErroneousUnsafe();
+			checkState(persistFuture.isDone(), "TODO support multiple pending persist requests (multiple ongoing checkpoints?)");
+			if (persistFuture.isDone()) {
+				persistFuture = new CompletableFuture<>();
+			}
+			add(PERSIST_MARKER);
+
+			return persistFuture;
+		}
+
+		public synchronized void checkErroneous() {
+			checkErroneousUnsafe();
+		}
+
+		@Override
+		public void run() {
+			try {
+				while (running) {
+					write(get());
+				}
+			}
+			catch (Throwable t) {
+				synchronized (this) {
+					if (running) {
+						asyncException = t;
+					}
+					if (!persistFuture.isDone()) {
+						persistFuture.completeExceptionally(t);
+					}
+				}
+				LOG.error("unhandled exception in the Writer", t);
+			}
+		}
+
+		private void write(Buffer buffer) throws IOException {
+			try {
+				int offset = buffer.getMemorySegmentOffset();
+				MemorySegment segment = buffer.getMemorySegment();
+				int numBytes = buffer.getSize();
+
+				if (readBuffer.length < numBytes) {
+					readBuffer = new byte[numBytes];
+				}
+				segment.get(offset, readBuffer, 0, numBytes);
+				currentOutputStream.write(readBuffer, 0, numBytes);
+			}
+			finally {
+				buffer.recycleBuffer();
+			}
+		}
+
+		private synchronized Buffer get() throws InterruptedException, IOException {
+			while (handover.isEmpty()) {
+				wait();
+			}
+			Buffer buffer = handover.poll();
+			if (buffer == PERSIST_MARKER) {
+				currentOutputStream.closeForCommit().commit();
+				openNewOutputStream();
+				persistFuture.complete(null);
+				return get();
+			}
+
+			return buffer;
+		}
+
+		@Override
+		public void close() throws InterruptedException, IOException {
+			try {
+				checkErroneous();
+				running = false;
+				interrupt();
+				join();
+			}
+			finally {
+				currentOutputStream.close();
+			}
+		}
+
+		private void openNewOutputStream() throws IOException {
+			currentOutputStream = recoverableWriter.open(assemblePartFilePath());
+		}
+
+		private Path assemblePartFilePath() {
+			return new Path(recoverableWriterBasePath, "part-file." + partId++);
```

Comment:
I guess all the tasks parallelism for one vertex would reference the same file path in practice? If so we should add one more level child path, e.g. execution id to distinguish.

Reply:
Yes makes sense, I added executionattempt id, not sure if this is exactly the id that you had in mind.

Reply:
exactly